### ISSUE TITLE
Reproduce results

### ISSUE THREAD

**Issue by Trainingzy** [wvangansbeke/Unsupervised-Semantic-Segmentation#16]
Thanks for your great work!

I run your code with ImageNet pretrained model and kmeans clustering evaluation. Got 42.2 mIOU (44.2 in your paper). It seems that you do not fix a seed for training. So how do you get the results? Average over multiple training results?

**Comment by Trainingzy** [comment]
The setting is: ImageNet supervised pretrained model, supervised saliency map, and kmeans clustering evaluation. Your report result is 44.2.

I further run your code for 4 times. The results are 42.2, 41.4, 43.9, 44.4. The average result is 43.0... Still smaller than the reported 44.2.

**Comment by wvangansbeke** [comment]
Hi @Trainingzy 

Please check the code [here](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/main/segmentation/kmeans.py#L78). As you can see the result are averaged over 5 seeds. This should give you 44 mIoU.  There might be some small differences due to the different versions of packages, but I doubt that.

**Comment by Trainingzy** [comment]
Hi @wvangansbeke, 

I know you average the results with 5 seeds in the kmeans evaluation process.  However, my four results (42.2, 41.4, 43.9, 44.4) are from 4 training models, and each of the models is evaluated directly by run your kmeans code, which is the results over 5 seeds.

It means that I run your method for 4 times and only one time I got a similar results.

**Comment by wvangansbeke** [comment]
Hi @Trainingzy,

Oh I see. Thanks for pointing this out. You are talking about the seed during training and not during Kmeans. Well there are always some differences between multiple runs. If you're only 1% off after running Kmeans, I don't think the training seed is that important (and it shouldn't be!). I also ran it two times and got almost exactly the same results. Keep in mind that different environments, different runs and training settings can produce slightly different results. I wouldn't lose sleep over this. More importantly, remember that this method does not use semantic annotations, so small variations when running Kmeans on top are expected! After all, you can cluster a dataset in many different ways which all can be correct. Unsurprisingly, some will align better with the predefined classes than others. If you want less variation, simply train a linear classifier on top. Code is provided.

However, based on your numbers, I argue that the results are reproducible. A small change of the hyperparameters or simply increasing `num_seeds` in `kmeans.py` will reduce the variance. All your numbers look similar to me though. Since you got the same value and also even a better one, I believe it should be fine. The shared model combined with the training/evaluation code (without any tricks) will give you exactly 44.2mIoU.

**Comment by Trainingzy** [comment]
Reasonable~ Thanks
